ui: Separate the invalidation time from the timeout

BramGruneir opened the pull request:
The invalidation time for cashed data reducers was used for both invalidation and as the api timeout. This change turns that into two different optional values.

Also included in this change:

- Set all debug pages to use KeyedCacheDataReducers instead of CachedDataReducers which fixes a number of bugs in which parts of the debug pages
might not update.
- Set the invalidation period for all debug pages to 0, so they are already reloaded intead of using cached values.
- Set the timeout for all debug pages to be 1m or greater (depending on the expected lenght of the request).

Fixes: #20634, #19920, #20683

Release note: None

BramGruneir committed:
ui: CachedDataReducer separate the invalidation time from the timeout

The invalidation time for cashed data reducers was used for both invalidation and as the api timeout. This change turns that into two different optional values.

Also included in this change:

- Set all debug pages to use KeyedCacheDataReducers instead of CachedDataReducers which fixes a number of bugs in which parts of the debug pages
might not update.
- Set the invalidation period for all debug pages to 0, so they are already reloaded intead of using cached values.
- Set the timeout for all debug pages to be 1m or greater (depending on the expected lenght of the request).

Fixes: #20634, #19920, #20683

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20828)
<!-- Reviewable:end -->

BramGruneir commented:
So I had the prettier extension installed and it kind of messed with a lot of the formatting.  Most for the better.  But if there's anything that seems worse, let me know and I'll got back and fix it.

vilterp commented on pkg/ui/src/redux/apiReducers.ts:
```diff
@@ -54,50 +90,138 @@ export function generateTableID(db: string, table: string) {
+const queryPlanReducerObj = new CachedDataReducer(
+  api.getQueryPlan,
+  "queryPlan",
+);
```
this is nowhere near the line limit… not sure why Prettier wrapped it…

vilterp commented on pkg/ui/src/views/reports/containers/range/index.tsx:
```diff
@@ -158,20 +171,23 @@ class Range extends React.Component<RangeProps, {}> {
+    lastError: state.cachedData.range[rangeID] && state.cachedData.range[rangeID].lastError,
+    allocatorLastError: state.cachedData.allocatorRange[rangeID] && state.cachedData.allocatorRange[rangeID].lastError,
+    rangeLogLastError: state.cachedData.rangeLog[rangeID] && state.cachedData.rangeLog[rangeID].lastError,
+    rangeLogInFlight: state.cachedData.rangeLog[rangeID] && state.cachedData.rangeLog[rangeID].inFlight,
```
I'd prefer we pass the `CachedDataReducer` state to each, as opposed to three variables for each (the data, `lastError`, and `inFlight`)

vilterp reviewed (approved):
Looks good overall. Again, I prefer passing in the reducer state to splitting it out into `loading` and `lastError` props, but we can do that in a later pass…

This is a step forward and we should get it in 👍

vilterp commented on pkg/ui/src/views/reports/containers/range/index.tsx:
```diff
@@ -82,38 +89,44 @@ class Range extends React.Component<RangeProps, {}> {
+          errorText={`No results found, perhaps r${
+            this.props.params[rangeIDAttr]
+            } doesn't exist.`}
+          rangeResponse={range}
```
kind of prefer the old wrapping here… maybe we should just go all in on prettier and not keep changing it's output though… idk

BramGruneir commented:
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/ui/src/redux/apiReducers.ts, line 151 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fEbTqV80f9et4bCHs-r1-151:-L0fMggTSqOZa_d6f5_S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe3969d3ae3de7f51e96b7eeef1196cc05b3f784/pkg/ui/src/redux/apiReducers.ts#L151)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

this is nowhere near the line limit… not sure why Prettier wrapped it…
</blockquote></details>

Done.

---

*[pkg/ui/src/views/reports/containers/range/index.tsx, line 190 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fEbTqV80f9et4bCHx-r1-190:-L0fU6S_2Mac8cneuDan:bywzx4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe3969d3ae3de7f51e96b7eeef1196cc05b3f784/pkg/ui/src/views/reports/containers/range/index.tsx#L190)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

I'd prefer we pass the `CachedDataReducer` state to each, as opposed to three variables for each (the data, `lastError`, and `inFlight`)
</blockquote></details>

Done... I just hope I caught all the null checks.

I do want to revisit all of these sub-page and give them proper loading and error pages.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/ui/src/views/reports/containers/range/index.tsx, line 127 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fEbTqV80f9et4bCHx-r1-127:-L0fYdCdcyIaQAwGbhJI:bf5zgty) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe3969d3ae3de7f51e96b7eeef1196cc05b3f784/pkg/ui/src/views/reports/containers/range/index.tsx#L127)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

kind of prefer the old wrapping here… maybe we should just go all in on prettier and not keep changing it's output though… idk
</blockquote></details>

I fixed the line wrapping of the error text, but the brackets around the jsx component is our style.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Reviewed 1 of 5 files at r2.
Review status: 1 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/ui/src/views/reports/containers/range/allocator.tsx, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fXXFQ6gQ3gJdTWrPA:-L0fXXFQ6gQ3gJdTWrPB:b-ythx93) ([raw file](https://github.com/cockroachdb/cockroach/blob/bad85d51d27ac4a347c28c21950135c29d8c2628/pkg/ui/src/views/reports/containers/range/allocator.tsx#L36)):*
> ```XML
>     return (
>       <div>
>         <h2>Simulated Allocator Output (from n{allocatorRangeResponse.data.node_id.toString()})</h2>
> ```

won't `.data` be null/undefined if the data hasn't loaded yet? Is that checked somewhere else?

---

*[pkg/ui/src/views/reports/containers/range/connectionsTable.tsx, line 15 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fXpt2JX7zIR0F6vMn:-L0fXpt2JX7zIR0F6vMo:bohn5j0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bad85d51d27ac4a347c28c21950135c29d8c2628/pkg/ui/src/views/reports/containers/range/connectionsTable.tsx#L15)):*
> ```XML
>   const { range } = props;
>   if (!range || _.isNil(range.data)) {
>     return null;
> ```

put a spinner here if you feel like it…

---

*[pkg/ui/src/views/reports/containers/range/logTable.tsx, line 85 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fY5Ppv90ipwoN8sEn:-L0fY5Ppv90ipwoN8sEo:b-ds2gr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/bad85d51d27ac4a347c28c21950135c29d8c2628/pkg/ui/src/views/reports/containers/range/logTable.tsx#L85)):*
> ```XML
>     const { log } = this.props;
> 
>     if (log && !_.isEmpty(log.lastError)) {
> ```

we've really got to wrap up these checks into a common `DataProvider` component or something…

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
ok, good to merge?

---

Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/ui/src/views/reports/containers/range/allocator.tsx, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fXXFQ6gQ3gJdTWrPA:-L0fcinBFA94Wl75oyPr:b-a2pcu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/bad85d51d27ac4a347c28c21950135c29d8c2628/pkg/ui/src/views/reports/containers/range/allocator.tsx#L36)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

won't `.data` be null/undefined if the data hasn't loaded yet? Is that checked somewhere else?
</blockquote></details>

ick, yeah, added a spinner too.

---

*[pkg/ui/src/views/reports/containers/range/connectionsTable.tsx, line 15 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fXpt2JX7zIR0F6vMn:-L0fchCZwCSgDc_P9eJc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bad85d51d27ac4a347c28c21950135c29d8c2628/pkg/ui/src/views/reports/containers/range/connectionsTable.tsx#L15)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

put a spinner here if you feel like it…
</blockquote></details>

Done.

---

*[pkg/ui/src/views/reports/containers/range/logTable.tsx, line 85 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fY5Ppv90ipwoN8sEn:-L0fclO0OmnIAx8ewSCX:bjz6qg0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bad85d51d27ac4a347c28c21950135c29d8c2628/pkg/ui/src/views/reports/containers/range/logTable.tsx#L85)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

we've really got to wrap up these checks into a common `DataProvider` component or something…
</blockquote></details>

I agree entirely.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-:-L0fdSpUPZp-TH32xhDs:b-aat0ea)*
<!-- Sent from Reviewable.io -->

vilterp commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 7 files at r1, 3 of 5 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/ui/src/views/reports/containers/range/allocator.tsx, line 12 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0ffjPWKPPSHXtlq9Zv:-L0ffjPWKPPSHXtlq9Zw:bi85xd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0d0784774763a5af1370c9627bdb19053169bc9/pkg/ui/src/views/reports/containers/range/allocator.tsx#L12)):*
> ```XML
> 
> interface AllocatorOutputProps {
>   allocatorRangeResponse: CachedDataReducerState<protos.cockroach.server.serverpb.AllocatorRangeResponse>;
> ```

nit: it's not the response if it's the whole `CachedDataReducerState`…

(not sure how worried I am about nits like this though, when what we really need is a `DataProvider` component or something)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-:-L0fg7ujuGbX9fJ6bsAW:bnfp4nl)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Fixed the nit, and made a small fix to the loading screen to be loading when the cacheddatareducers were null.

---

Review status: 5 of 9 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/ui/src/views/reports/containers/range/allocator.tsx, line 12 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0ffjPWKPPSHXtlq9Zv:-L0fiuWt3PWukbwmYiCj:b67ov77) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0d0784774763a5af1370c9627bdb19053169bc9/pkg/ui/src/views/reports/containers/range/allocator.tsx#L12)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

nit: it's not the response if it's the whole `CachedDataReducerState`…

(not sure how worried I am about nits like this though, when what we really need is a `DataProvider` component or something)
</blockquote></details>


Renamed. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-:-L0fiy1nHt85E1b2mqTy:bry2d2o)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Cool. Ready to merge, I think. Thinking about a `<DataProvider .../>` component… This is a complicated topic; will have to discuss with @couchand when he comes back.

couchand commented on pkg/ui/src/views/reports/containers/range/logTable.tsx:
```diff
@@ -57,10 +61,15 @@ export default class LogTable extends React.Component<LogTableProps, {}> {
+        {this.renderLogInfoDescriptor(
+          "Updated Range Descriptor",
+          info.updated_desc,
+        )}
```
ugh I don't like that just one of these got multi-lined

couchand commented on pkg/ui/src/views/reports/containers/range/logTable.tsx:
```diff
@@ -71,55 +80,75 @@ export default class LogTable extends React.Component<LogTableProps, {}> {
-                <th className="log-table__cell log-table__cell--header">Other Range</th>
-                <th className="log-table__cell log-table__cell--header">Info</th>
+                <th className="log-table__cell log-table__cell--header">
+                  Timestamp
```
another bleh. oh well.

couchand commented on pkg/ui/src/redux/cachedDataReducer.ts:
```diff
@@ -52,11 +52,13 @@ export class CachedDataReducer<TRequest, TResponseMessage> {
     protected apiEndpoint: APIRequestFn<TRequest, TResponseMessage>,
     public actionNamespace: string,
     protected invalidationPeriod?: moment.Duration,
+    protected requestTimeout?: moment.Duration,
```
should this default to the `invalidationPeriod` to keep the old behavior maybe?  or was that design flawed anyway, and no default is better?

couchand reviewed (commented):
@BramGruneir thanks for fixing this.

I know this PR's already merged, left a few comments anyway.  None of them are particularly important.

In the future, please please please put formatting changes in a different commit from meaningful changes.  It makes code review much harder to have to read through all of this just to find the few things that changed.

BramGruneir commented:
So I removed prettier... ugh.  what a hassle that was.  See #21529 which fixes the rest of the issues from here.

---

Review status: 5 of 9 files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[pkg/ui/src/redux/apiReducers.ts, line 132 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fEbTqV80f9et4bCHs-r4-132:-L34dpmeOD7WG8tTKZzS:b-w9d3bq) ([raw file](https://github.com/cockroachdb/cockroach/blob/284315a6c9c7859297fc9a41b4baca0fb8c5b17e/pkg/ui/src/redux/apiReducers.ts#L132)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

ugh this is horrible.  prettier usually does a good job, but not here!
</blockquote></details>

Yeah, that was some uglyness.  fixed.

---

*[pkg/ui/src/redux/apiReducers.ts, line 146 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fEbTqV80f9et4bCHs-r4-146:-L34dw15Zl2crEeXglI4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/284315a6c9c7859297fc9a41b4baca0fb8c5b17e/pkg/ui/src/redux/apiReducers.ts#L146)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

really, prettier?
</blockquote></details>

Done.

---

*[pkg/ui/src/redux/apiReducers.ts, line 153 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fEbTqV80f9et4bCHs-r4-153:-L34dz3ZWnbJu3Wo7oU4:b-ect4d3) ([raw file](https://github.com/cockroachdb/cockroach/blob/284315a6c9c7859297fc9a41b4baca0fb8c5b17e/pkg/ui/src/redux/apiReducers.ts#L153)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

hmm, should that be `isNil` for clarity?
</blockquote></details>

In this case, it's already a string.  And an empty string is not nil.  Anyway, touched this as part of #21522.

---

*[pkg/ui/src/redux/apiReducers.ts, line 165 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fEbTqV80f9et4bCHs-r4-165:-L34eZku7TI2Ae_DMBqn:b-hkwdip) ([raw file](https://github.com/cockroachdb/cockroach/blob/284315a6c9c7859297fc9a41b4baca0fb8c5b17e/pkg/ui/src/redux/apiReducers.ts#L165)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Again, it's a string.  But I did some cleanup on this in #21522.  Let me know if you think this should be updated further.

---

*[pkg/ui/src/redux/cachedDataReducer.ts, line 61 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fEbTqV80f9et4bCHt-r4-61:-L34ehdKLyaDNxB85rTP:b-6xsydv) ([raw file](https://github.com/cockroachdb/cockroach/blob/284315a6c9c7859297fc9a41b4baca0fb8c5b17e/pkg/ui/src/redux/cachedDataReducer.ts#L61)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

should this default to the `invalidationPeriod` to keep the old behavior maybe?  or was that design flawed anyway, and no default is better?
</blockquote></details>

That old design was flawed.  And I think the conflation of the two concepts was based on confusion.

---

*[pkg/ui/src/redux/cachedDataReducer.ts, line 227 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fEbTqV80f9et4bCHt-r4-227:-L34dT5_rGFzGpg2QHxQ:byaajf9) ([raw file](https://github.com/cockroachdb/cockroach/blob/284315a6c9c7859297fc9a41b4baca0fb8c5b17e/pkg/ui/src/redux/cachedDataReducer.ts#L227)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

why does this line have two parameters?  come on, prettier!
</blockquote></details>

This one I fixed instinctively while looking over the code in #21522.

---

*[pkg/ui/src/views/reports/containers/certificates/index.tsx, line 189 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fEbTqV80f9et4bCHu-r4-189:-L34f2Upv3-ZRbxZU4aP:b67eueb) ([raw file](https://github.com/cockroachdb/cockroach/blob/284315a6c9c7859297fc9a41b4baca0fb8c5b17e/pkg/ui/src/views/reports/containers/certificates/index.tsx#L189)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

It's really too bad that these changes are so far from the other side.  Seems like there's an opportunity for a refactor here.  I suppose the limitation there is that the `KeyedCachedDataReducer` and the selector here are getting the key from different places...
</blockquote></details>

Yeah, that was the cause of the problem fixed in #21522.  I'd rather the key selection happens once, not in two different places.  I've made an adjustment to the cert page to also use the key selector from the apiReducers.

---

*[pkg/ui/src/views/reports/containers/range/allocator.tsx, line 44 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fU3kJtrOEnjHZlOAv-r4-44:-L34iEjREl4qrxX5pwAv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/284315a6c9c7859297fc9a41b4baca0fb8c5b17e/pkg/ui/src/views/reports/containers/range/allocator.tsx#L44)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

no space needed before curly brace.
</blockquote></details>

Done.

---

*[pkg/ui/src/views/reports/containers/range/connectionsTable.tsx, line 18 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fU3kJtrOEnjHZlOAw-r4-18:-L34iRyKA-6NIuT5dK5h:buzm4kx) ([raw file](https://github.com/cockroachdb/cockroach/blob/284315a6c9c7859297fc9a41b4baca0fb8c5b17e/pkg/ui/src/views/reports/containers/range/connectionsTable.tsx#L18)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

doesn't this kind of defeat the purpose of the loading state?
</blockquote></details>

ick, you're right.  Fixed.

---

*[pkg/ui/src/views/reports/containers/range/logTable.tsx, line 72 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fEbTqV80f9et4bCHy-r4-72:-L34jR-Luy2muWymcWd0:b-nz12qn) ([raw file](https://github.com/cockroachdb/cockroach/blob/284315a6c9c7859297fc9a41b4baca0fb8c5b17e/pkg/ui/src/views/reports/containers/range/logTable.tsx#L72)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

ugh I don't like that just one of these got multi-lined
</blockquote></details>

ugh is right... fixed.

---

*[pkg/ui/src/views/reports/containers/range/logTable.tsx, line 114 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-L0fEbTqV80f9et4bCHy-r4-114:-L34jljEkm1HxWctwbTd:bq03krq) ([raw file](https://github.com/cockroachdb/cockroach/blob/284315a6c9c7859297fc9a41b4baca0fb8c5b17e/pkg/ui/src/views/reports/containers/range/logTable.tsx#L114)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

another bleh. oh well.
</blockquote></details>

restored!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20828#-:-L34dWl17s722uoL9Y8S:b-vqs3f2)*
<!-- Sent from Reviewable.io -->
